src/Std/Sat/AIG/Basic.lean

```diff
@@ -0,0 +1,475 @@
+    TernaryInput aig2 :=
+  { input with discr := input.discr.cast h, lhs := input.lhs.cast h, rhs := input.rhs.cast h }
+
+/--
+An entrypoint into an `AIG`. This can be used to evaluate a circuit, starting at a certain node,
+with `AIG.denote` or to construct bigger circuits on top of this specific node.
+-/
+structure Entrypoint (α : Type) [DecidableEq α] [Hashable α] where
+  /--
+  The AIG that we are in.
+  -/
+  aig : AIG α
+  /--
+  The reference to the node in `aig` that this `Entrypoint` targets.
+  -/
+  ref : Ref aig
+
+/--
+Transform an `Entrypoint` into a graphviz string. Useful for debugging purposes.
+-/
+def toGraphviz {α : Type} [DecidableEq α] [ToString α] [Hashable α] (entry : Entrypoint α)
+    : String :=
+  let ⟨⟨decls, _, hinv⟩, ⟨idx, h⟩⟩ := entry
+  let (dag, s) := go "" decls hinv idx h |>.run .empty
+  let nodes := s.fold (fun x y ↦ x ++ toGraphvizString decls y) ""
+  "Digraph AIG {" ++ nodes ++ dag ++ "}"
+where
+  go {α : Type} [DecidableEq α] [ToString α] [Hashable α] (acc : String) (decls : Array (Decl α))
+      (hinv : IsDag α decls) (idx : Nat) (hidx : idx < decls.size)
+        : StateM (HashSet (Fin decls.size)) String := do
+    let fidx : Fin decls.size := Fin.mk idx hidx
+    if (← get).contains fidx then
+      return acc
+    modify (fun s ↦ s.insert fidx)
+    match elem : decls[idx] with
+    | Decl.const _ => return acc
+    | Decl.atom _ => return acc
+    | Decl.gate lidx ridx linv rinv =>
+      let curr := s!"{idx} -> {lidx}{invEdgeStyle linv}; {idx} -> {ridx}{invEdgeStyle rinv};"
+      let hlr := hinv idx lidx ridx linv rinv hidx elem
+      let laig ← go (acc ++ curr) decls hinv lidx (by omega)
+      go laig decls hinv ridx (by omega)
+  invEdgeStyle (isInv : Bool) : String :=
+    if isInv then " [color=red]" else " [color=blue]"
+  toGraphvizString {α : Type} [DecidableEq α] [ToString α] [Hashable α] (decls : Array (Decl α))
+      (idx : Fin decls.size) : String :=
+    match decls[idx] with
+    | Decl.const b => s!"{idx} [label=\"{b}\", shape=box];"
+    | Decl.atom i => s!"{idx} [label=\"{i}\", shape=doublecircle];"
+    | Decl.gate _ _ _ _ => s!"{idx} [label=\"{idx} ∧\",shape=trapezium];"
+
+/--
+A stream of references into `aig`. This is the `AIG` analog of `BitVec`.
+-/
+structure RefStream (aig : AIG α) (length : Nat) where
+  refs : Array Nat
+  hlen : refs.size = length
+  hrefs : ∀ (h : i < length), refs[i] < aig.decls.size
+
+/--
+A sequence of references bundled with their AIG. In some sense a generalization of `Entrypoint`.
+-/
+structure RefStreamEntry (α : Type) [DecidableEq α] [Hashable α] [DecidableEq α] (length : Nat) where
+  aig : AIG α
+  stream : RefStream aig length
+
+structure ShiftTarget (aig : AIG α) (w : Nat) where
+  stream : AIG.RefStream aig w
+  distance : Nat
+
+structure ArbitraryShiftTarget (aig : AIG α) (m : Nat) where
+  n : Nat
+  target : AIG.RefStream aig m
+  distance : AIG.RefStream aig n
+
+structure ExtendTarget (aig : AIG α) (newWidth : Nat) where
```

Comment:
You're defining a lot of structures in this file, some of which aren't used in this PR. Are they used later in the development? For some of them I'm also wondering whether bundling this data together is really worth it.

Reply:
Yes, these are all used in the bit blasting part. They are mostly bundled because I need my functions to be a lawful operator which only takes one argument.

Reply:
Ah, I see, thanks. No need to change anything for now.

Reply:
But I think we should nevertheless have doc-strings, even if just "an X bundled with Y".